doc/docgen.rst

```diff
@@ -229,6 +229,146 @@ Output::
+
+   A. non-qualified:
+
+      1) specifying parameters names::
+
+          `foo(a, b)`_
+
+      2) specifying parameters types::
+
+          `foo(int, float)`_
+
+      3) specifying both names and types::
+
+          `foo(a: int, b: float)`_
+
+      4) output parameter can also be specified if you wish::
+
+          `foo(a: int, b: float): string`_
+
+   B. qualified: all 4 options above are valid.
+      Particularly you can use the full format::
+
+        `proc foo(a: int, b: float): string`_
+
+   **A limitation**:
+   The parameters of a nested routine type can be specified only with types
+   (without parameter names, see form A.2 above). E.g. for this signature:
+
+   .. code:: Nim
+
+      proc binarySearch*[T, K](a: openArray[T]; key: K;
+                               cmp: proc (x: T; y: K): int {.closure.}): int
+                                          ~~    ~~   ~~~~~
+
+   you cannot use names underlined by `~~` so it must be referenced with
+   ``cmp: proc(T, K)``. Hence these forms are valid::
+
+     `binarySearch(a: openArray[T], key: K, cmp: proc(T, K))`_
+     `binarySearch(openArray[T], K, proc(T, K))`_
+     `binarySearch(a, key, cmp)`_
+
+Generic parameters can also be used. All in all, this long form will be
+recognized fine::
+
+  `proc binarySearch*[T, K](a: openArray[T], key: K, cmp: proc(T, K)): int`_
+
+.. Tip:: Avoid extraneous information by using one of these 2 short forms
+   in text::
+
+     binarySearch_
+     `proc binarySearch(a, key, cmp)`_
+
+   If you use a short form and have an ambiguity then just add some
+   additional info.
+
+.. Warning:: An ambiguity in resolving documentation links may arise because of:
+
+   1. clash with other RST anchors
+      * manually setup anchors
+      * automatically set up, e.g. section names
+   2. collision with other Nim symbols.
+      Because in Nim `proc` and `iterator` belong to different namespaces.
+      Note that qualified references are useful in this case: just use the
+      qualified form like `\`proc foo\`_`:literal: or
+      `\`iterator foo\`_`:literal:.
+
+   Any ambiguity is always reported with Nim compiler warnings and an anchor
+   with higher priority is selected. Manual anchors have highest
+   priority, then go automatic RST anchors; then Nim-generated anchors
+   (while procs have higher priority than other Nim symbol kinds).
+
+.. Note:: A bit special case is operators
+   (as their signature is also defined with `\``):
+
+   .. code:: Nim
+
+      func `[]`*[T](x: openArray[T]): T
+      func `$`(x: MyType): string
+
+   A short form works without additional backticks::
+
+     `[]`_
+     `$`_
+
+   However for fully-qualified reference copy-pasting backticks (`) into other
+   backticks will not work in our RST parser (because we use Markdown-like
+   inline markup rules). In the case of `$` it's allowed to delete backticks,
+   but in case of `[]` one needs to keep backticks and escape them with
```

Comment:
that's a bit nasty, can we avoid this?
this should work and not be ambiguous:
```nim
`func []`_ or, in detail:
`func [][T](x: openArray[T]): T`_
```

Reply:
yes, at first glance I don't see any principal obstacle. It may well be a bug in my working copy :-)

Reply:
it's working now.